File: LeetCode/src/kadane/MaximumSubarray.java
package kadane;

public class MaximumSubarray {
    public int maxSubArray(int[] nums) {
        int maxSum = nums[0];
        int curSum = nums[0];

        for (int i = 1; i < nums.length; i++) {
            int num = nums[i];
            curSum = Math.max(num, curSum + num);
            maxSum = Math.max(maxSum, curSum);
        }

        return maxSum;
    }

    public static void main(String[] args) {
        MaximumSubarray solution = new MaximumSubarray();

        // Test Case 1
        int[] nums1 = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
        int expected1 = 6;
        int result1 = solution.maxSubArray(nums1);
        System.out.println("Test Case 1:");
        System.out.println("Expected: " + expected1);
        System.out.println("Result: " + result1);

        // Test Case 2
        int[] nums2 = {1};
        int expected2 = 1;
        int result2 = solution.maxSubArray(nums2);
        System.out.println("Test Case 2:");
        System.out.println("Expected: " + expected2);
        System.out.println("Result: " + result2);

        // Test Case 3
        int[] nums3 = {5, 4, -1, 7, 8};
        int expected3 = 23;
        int result3 = solution.maxSubArray(nums3);
        System.out.println("Test Case 3:");
        System.out.println("Expected: " + expected3);
        System.out.println("Result: " + result3);
    }
}

File: LeetCode_Python/dp/kadane/MaximumSubarray.py
class MaximumSubarray:
    def maxSubArray(self, nums: list[int]) -> int:
        cur_sum = max_sum = nums[0]

        for n in nums[1:]:
            cur_sum = max(n, n + cur_sum)
            max_sum = max(max_sum, cur_sum)

        return max_sum


if __name__ == '__main__':
    solution = MaximumSubarray()

    nums1 = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    print("Expected Output:", 6)
    print("Actual Output:", solution.maxSubArray(nums1))

    nums2 = [1]
    print("Expected Output:", 1)
    print("Actual Output:", solution.maxSubArray(nums2))

    nums3 = [-3, -2, -1, -4]
    print("Expected Output:", -1)
    print("Actual Output:", solution.maxSubArray(nums3))
